### PR TITLE
feat: Support Line Column in file pickers

### LINKED ISSUE #2650: Add an ability to open direct location
**Is your feature request related to a problem? Please describe.**
I usually deal with direct location output from the compiler which is a path to the file plus the location marker. e.g.

```
my/directory/file.rs:24:4
```

**Describe the solution you'd like**
I want to be able to paste this directly to the telescope, press enter, and appear in the direct symbol location.

> P.S. This is possible in vs code with no additional setup for example

**Describe alternatives you've considered**
1. Open telescope search
2. Observe nothing found
3. Remove paths 
4. type :24<CR> to go to the line 

**Additional context**
This should be primarily an input feature — strip out everything that looks like location, search, and when open file focus specific symbol. That's why I think it should be a part of the telescope core and not be solved via plugins. 

If you are interested I can open PR for this

### PR DESCRIPTION
**Pull request by dmtrKovalenko**
## Type of change

- New feature (non-breaking change which adds functionality)

## Description

Fixing the original issue: When selecting files in the telescope impossible to copy paths from the compiler, you need to clear the location from the path.

Fixes #2650

This fixes the problem by stripping out the location from the path (important: only for file pickers). And using the location to set the cursor when opening the file.

There is a new option `files_picker` for a `Picker` that needs to be set to true for all file pickers, for now, it basically enables location stripping which makes no sense in another type of pickers.

Watch the video:

https://github.com/nvim-telescope/telescope.nvim/assets/16926049/1ee3b507-4af9-4d0f-abe2-3c67254dfcd9

## Implementation notice

I didn't find another way to put the location stripping because it must happen right after the input and make sure that if the location is set the prompt = input without location. But after we calculated that during selection we can set `entry.lnum` and `entry.col` to the value and reuse all the existing functionality to jump to the cursor

## How it is different from https://github.com/nvim-telescope/telescope.nvim/pull/2312

The PR seems to be abandoned but I need this functionality so much so I updated and changed it myself, made the API closer to the current entry row/linenr selection state which is used in live_grep, implemented support for the column number, added automated tests, and highlighted the line and column in the previewer.

## How Has This Been Tested?

An added test suite that covers the algorithm of stripping the location from the path to the `utils_spec.lua` 

Manual test:

1. Open telescope to find files
2. Enter a:12:1
3. Observer that files with a in the name found
4. Open the file and observe that you are in the :12:1 position

**Configuration**:
* Neovim version (nvim --version):

NVIM v0.9.4
Build type: Release
LuaJIT 2.1.1699180677
* Operating system and version: MacOS 13

# Checklist:

- [x] My code follows the style guidelines of this project (stylua)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (lua annotations)

### CODE REVIEW

File: lua/telescope/pickers.lua
Comment (dmtrKovalenko) [inline lua/telescope/pickers.lua:337]:
```suggestion
      obj.preview_title == nil
      or #obj.all_previewers > 1
      or opts.resumed_picker and opts.fix_preview_title ~= true
```

File: lua/telescope/pickers.lua
Comment (Conni2461) [inline lua/telescope/pickers.lua:648]: this doesnt really address this comment: https://github.com/nvim-telescope/telescope.nvim/pull/2312#issuecomment-1479645530

i agree with tj here, that i dont really want builtin specific code in `pickers.lua`, because i dont think we should couple specific builtin code with the framework.

The framework should rather expose callbacks that allow us to do these things.

why can't we use `_on_input_filter_cb` its used in `_get_next_filtered_prompt` and we can then return the new prompt with `{ prompt = "asdf" }`

File: lua/telescope/pickers.lua
Comment (Conni2461) [inline lua/telescope/pickers.lua:1074]: not sure we should inject that here into the entry, i'd rather we just use that prompt_location in the actions, similar to what you did with the previewer, because with this right now you could have just used the `entry.lnum` and `entry.col` in previewer code, fyi

Why do you do it like this?

Comment (dmtrKovalenko) [reply lua/telescope/pickers.lua:1074]: To reuse the code of putting the cursor to the specific location, thus we will need to duplicate them but I’m find with that either

File: lua/telescope/pickers.lua
Comment (dmtrKovalenko) [reply lua/telescope/pickers.lua:648]: I can look into that as well

Comment (dmtrKovalenko) [reply lua/telescope/pickers.lua:648]: Hey @Conni2461 after a bit of investigation I realised why it would be a bad idea to put it into some callback — the biggest problem is the global state that we need to mutate and when the new picker is open we need to clear this state (or at least reset on reopening). I don't like an idea of stacking callbacks (we need to keep the functionality) and mutating global state inside of the stacked callbacks.

My proposal would be different. Define a new callback — `get_prompt_metadata` that will be executed in the picker and get the metadata from the prompt, then use this metadata to set the entry.lnum and entry.col on selection and also push this into the previewer.state. 

Thus we will never end up in a way when the global state is mutated. But every file-related picker will need to provide this `get_prompt_metadata` to achieve location parsing. 

If you are ok with this design please let me know — waiting for the green light.

Comment (Conni2461) [reply lua/telescope/pickers.lua:648]: hmm, let me think about it. I am also not that happy with a new kinda specific callback (cant think of another reason why we would need that)

Comment (dmtrKovalenko) [reply lua/telescope/pickers.lua:648]: I'll just put 3 things that are requirements for this feature (in my opinion): 

1. You need to mutate the prompt when the location is valid
2. You need to save the location and if it global state clear it for non supported picker
3. In my ideal world all of my pickers even external https://github.com/nvim-telescope/telescope-frecency.nvim will support this feature, or for instance I want to implement search by sapling files would support this feature, so having it as one flag option on the picker is preferred

Comment (Conni2461) [reply lua/telescope/pickers.lua:648]: please be patient. Its been 3 days (, i reviewed it, i started looking into different implementations but i am also really busy. This project is my free time and i which is generally limited to weekends.

I'll merge a solution for this feature, but i need to think about this a couple more days. I'll not rush a solution because the majority of vim users tend to run the default branch, which makes breaking changes a pain.

I think you can run your fork a couple more days because there werent any mayor commits on master anyway.

Comment (dmtrKovalenko) [reply lua/telescope/pickers.lua:648]: Sure, happy to understand that it is not lost.

Comment (Conni2461) [reply lua/telescope/pickers.lua:648]: i think i have a implementation i slighly prefer: https://github.com/nvim-telescope/telescope.nvim/commit/64fc6c58f984b17ae243da90a75f3bafd23227df

i am not entirely happy and we need some util functions so people can easily set the callback but i think that makes it less coupled inside the picker.

so its not done, but i also havent spend more than 30minutes on it.

Its pretty likely that i dont have time to look at this prior to saturday, just a heads up

Comment (dmtrKovalenko) [reply lua/telescope/pickers.lua:648]: Thank you for doing that, what confuses me in this is that the “filter_input_cb” sets totally global state which is super unclear side effect as for me. Just a thought, happy to help with any implementation you choose.

just let me know what I can do, thank you for your help on this!!

Comment (Conni2461) [reply lua/telescope/pickers.lua:648]: yeah what i dont like about your implementation is that picker now does things that are totally only related to file specific things.

the picker itself, ui, sorting, etc, has nothing to do with files it operates on lists.

> telescope.nvim is a highly extendable fuzzy finder over lists. Built on the latest awesome features from neovim core. Telescope is centered around modularity, allowing for easy customization.

its also not a global global thing. i was thinking about introducing this local_state, which gets cleaned up after shutdown and allows users to save information, that is then accessible in the different stages of the picker (preview and actions). i am not happy that the default action/previewer have knowledge about this key/value pair, thats why i think its not done yet, but i think its a step up, because the picker no longer is responsible for it.

**Review by xiantang — APPROVED** [review]

**Comment by dmtrKovalenko** [comment]
@Conni2461 it is already a month — could I somehow help to get this merged? If you have any ideas of how I can improve the implementation please briefly write them here and I'll try to add them. 

Or do you want to simply move forward with your version https://github.com/nvim-telescope/telescope.nvim/commit/64fc6c58f984b17ae243da90a75f3bafd23227df?

**Comment by Conni2461** [comment]
Okay lets merge this as an "experimental" feature ... I still wanna mention that I dont like the application Interface, and i wanna 100% rework it but this isn't possible without some mayor effort with some kinda telescope hooking system which takes too long right now, conceptional and then the actual implementation of that. (Problems summed up in this comment thread: https://github.com/nvim-telescope/telescope.nvim/pull/2791#discussion_r1399037013 and in the original proposal https://github.com/nvim-telescope/telescope.nvim/pull/2312). 

TLDR: telescope core has nothing to do with files and having the core system be tied to this is a system design decision i dont like.

I need a couple of changes that bascially mark this as experimental after that i would merge this, even though it doesnt makes me happy. Comments regarding these changes, follow below

File: lua/telescope/builtin/__files.lua
Comment (Conni2461) [inline lua/telescope/builtin/__files.lua:387]: lets rename this to `__cursor_support` (i think, important thing is the `__` prefix)

File: lua/telescope/pickers.lua
Comment (Conni2461) [inline lua/telescope/pickers.lua:323]: same name on both sides, and prefix with `__`

File: lua/telescope/builtin/__files.lua
Comment (Conni2461) [reply lua/telescope/builtin/__files.lua:387]: or `__location_input` dont care.

File: lua/telescope/previewers/buffer_previewer.lua
Comment (Conni2461) [inline lua/telescope/previewers/buffer_previewer.lua:491]: we inject `entry` already with `lnum` and `col`. why not just use a picker that already has support for jumping to the line and highlighing the line. aka grep_previewer, isnt that good enough?

File: lua/telescope/utils.lua
Comment (Conni2461) [inline lua/telescope/utils.lua:581]: i dont think we need these comments

File: lua/telescope/utils.lua
Comment (Conni2461) [inline lua/telescope/utils.lua:573]: can we also prefix this function with `__`, i am happy to have it as a utils function because then its tested but i am not sure i want extension authors to use that function

File: lua/telescope/previewers/buffer_previewer.lua
Comment (dmtrKovalenko) [reply lua/telescope/previewers/buffer_previewer.lua:491]: I am here more afraid about potential issues and int the behavior because I might miss because the code is somewhat different and has some handlers that I am not really sure I understand 100%. 

If you 100% sure it will work we can try that out as well

**Comment by dmtrKovalenko** [comment]
@Conni2461 I addressed all the changes you requested except the grep previewer because I don't feel 100% sure it is working in all edge casees

**Comment by Conni2461** [comment]
I pushed 3 more commits, addressed https://github.com/nvim-telescope/telescope.nvim/pull/2791#discussion_r1438706606 and https://github.com/nvim-telescope/telescope.nvim/pull/2791#discussion_r1438707304

can you retest if everything still works for you? because i havent found more issues and it should be mergeable now.

**Comment by dmtrKovalenko** [comment]
I pulled the changes and everything works like a charm.  **Very big thanks for your help on this.**.

I have another small question related to this feature: would it be ok if I'll make a PR to https://github.com/nvim-telescope/telescope-frecency.nvim to add the `__locations_input` option and to make it also handle the locations? I recently tried to switch from the default pickers and would be happy to make it integration. Can we make this option public for some 3d party pickers?

**Comment by Conni2461** [comment]
I found a regression and fix that one today and pushed a fix so it actually be good to go, ill merge tomorrow morning.

Regarding frecency, i am not the current maintainer of that plugin so i am not sure delphinus wants this, this is up to him, but if they dont want this, you should be able to just enable this in your config. Because the opts are usually passed into `pickers.new` like this `pickers.new(opts, default_stuff)`. `__locations_input` doesnt have to be part of `default_stuff`, for it to work

I have not tested this but this should probably work
```lua
require("telescope").extensions.frecency.frecency{
  __locations_input = true,
  previewer = require("telescope.config").values.grep_previewer({}),

}
```

**Comment by dmtrKovalenko** [comment]
That's wonderful!

**Comment by Conni2461** [comment]
Thanks for the contribution :)

**Comment by xiantang** [comment]
great feature. I updated my telescope instantly.

**Comment by Ryel** [comment]
Thank you @dmtrKovalenko and @Conni2461 !!!

**Comment by josephcrawfordSRH** [comment]
How does one go about enabling this, is it in the release that is installed in lazyvim or would I have to install a specific pre-release version?  Highly interested in being able to jump to line numbers when jumping files, awesome work getting this in place!

**Comment by ctretyak** [comment]
how to turn it on? It doesn't work by default in 0.1.8 and cannot find anything in help

**Comment by jamestrew** [comment]
@ctretyak https://github.com/nvim-telescope/telescope.nvim/pull/2791#issuecomment-1962187594

**Comment by josephcrawfordSRH** [comment]
@jamestrew Is the same thing possible with telescope to say jump to function defenition so instead of doing `:21` to jump to the line can you also do `@methodName`?  This is something that I miss from my vscode days but still not enough for me to switch back :)

**Comment by jamestrew** [comment]
@josephcrawfordSRH 
see this https://github.com/nvim-telescope/telescope.nvim/issues/1042

I think adding `@` support to `find_files` could makes sense but there isn't any plan to do so currently.

I'm also little wary about turning `find_files` into a vscode/sublime style catch all finder.
We've added support for jumping to line/col of a file with this PR, I'm now seeing more requests for jumping directly to symbols with `@`.

If we add `@` support, next people will ask for `#` and `>` (workspace symbol & commands respectively) maybe :sweat_smile: 

I'm not saying these are bad ideas. But at some point, it's probably a separate new picker.

**Comment by josephcrawfordSRH** [comment]
@jamestrew thanks for getting back to me, sorry for my delay in response I was on PTO.  I completely understand your points above and was more curious if I was missing something but it sounds like I am not.  I do hope that something like this could be implemented in the future either in this picker or a different picker :)

**Pull request merged**
